codex/blockexchange/engine/pendingblocks.nim

```diff
@@ -37,88 +40,121 @@ type
     startTime*: int64
 
   PendingBlocksManager* = ref object of RootObj
-    blocks*: Table[Cid, BlockReq] # pending Block requests
+    blocks*: Table[BlockAddress, BlockReq] # pending Block requests
 
 proc updatePendingBlockGauge(p: PendingBlocksManager) =
   codex_block_exchange_pending_block_requests.set(p.blocks.len.int64)
 
 proc getWantHandle*(
     p: PendingBlocksManager,
-    cid: Cid,
+    address: BlockAddress,
     timeout = DefaultBlockTimeout,
     inFlight = false
 ): Future[Block] {.async.} =
   ## Add an event for a block
   ##
 
   try:
-    if cid notin p.blocks:
-      p.blocks[cid] = BlockReq(
+    if address notin p.blocks:
+      p.blocks[address] = BlockReq(
         handle: newFuture[Block]("pendingBlocks.getWantHandle"),
         inFlight: inFlight,
         startTime: getMonoTime().ticks)
 
-      trace "Adding pending future for block", cid, inFlight = p.blocks[cid].inFlight
+      trace "Adding pending future for block", address, inFlight = p.blocks[address].inFlight
 
     p.updatePendingBlockGauge()
-    return await p.blocks[cid].handle.wait(timeout)
+    return await p.blocks[address].handle.wait(timeout)
   except CancelledError as exc:
-    trace "Blocks cancelled", exc = exc.msg, cid
+    trace "Blocks cancelled", exc = exc.msg, address
     raise exc
   except CatchableError as exc:
     trace "Pending WANT failed or expired", exc = exc.msg
     # no need to cancel, it is already cancelled by wait()
     raise exc
   finally:
-    p.blocks.del(cid)
+    p.blocks.del(address)
     p.updatePendingBlockGauge()
 
-proc resolve*(p: PendingBlocksManager,
-              blocks: seq[Block]) =
+proc getWantHandle*(
+    p: PendingBlocksManager,
+    cid: Cid,
+    timeout = DefaultBlockTimeout,
+    inFlight = false
+): Future[Block] =
+  p.getWantHandle(BlockAddress.init(cid), timeout, inFlight)
+
+proc resolve*(
+  p: PendingBlocksManager,
+  blocksDelivery: seq[BlockDelivery]
+  ) {.gcsafe, raises: [].} =
   ## Resolve pending blocks
   ##
 
-  for blk in blocks:
-    # resolve any pending blocks
-    p.blocks.withValue(blk.cid, pending):
-      if not pending[].handle.completed:
-        trace "Resolving block", cid = blk.cid
-        pending[].handle.complete(blk)
+  for bd in blocksDelivery:
+    p.blocks.withValue(bd.address, blockReq):
+      trace "Resolving block", address = bd.address
+
+      if not blockReq.handle.finished:
```

Comment:
you want to use the `[]` for ref/pointer types, nim will try to infere the type, but it doesn't work always...

Reply:
But it works in this case, should I add `[]` anyway?

Reply:
Yeah, it doesn't work _always_ and it's also easy to miss that we're really handling a pointer when reading the code.